### PR TITLE
change mempool to use a non-recursive shared lock.

### PR DESCRIPTION
**Pull request by gandrewstone**
Change mempool to use a non-recursive shared lock.  Introduces the convention of using an underscore (_) to indicate non-locking versions of member functions.

Also fixes a small bug in argument handling in the qa rpc_tests.py program.

### CODE REVIEW

**Comment by sickpig** [comment]
@gandrewstone formatting error 
https://travis-ci.org/BitcoinUnlimited/BitcoinUnlimited/jobs/329127349#L750

guess it's time to merge @awemany's #631 

the amount of time a submitted PR get a failure due to formatting is starting to become pretty significant.

File: qa/pull-tester/rpc-tests.py [hunk 1]
```diff
@@ -143,7 +143,7 @@ def option_passed(option_without_dashes):
 bad_opt_str="Unrecognized option: %s"
 for o in opts | double_opts:
     if o.startswith('--'):
-        if o not in framework_opts + test_script_opts + private_double_opts:
+        if o.split("=")[0] not in framework_opts + test_script_opts + private_double_opts:
```
Comment (sickpig) [inline qa/pull-tester/rpc-tests.py:146]: this change is unrelated.

File: src/thinblock.cpp [hunk 2]
```diff
@@ -163,7 +163,7 @@ bool CThinBlock::process(CNode *pfrom, int nSizeThinBlock)
 
     {
         LOCK(cs_orphancache);
-        LOCK2(mempool.cs, cs_xval);
+        LOCK(cs_xval);
         int missingCount = 0;
```
Comment (ptschip) [inline src/thinblock.cpp:168]: I believe that taking out the mempool lock here breaks the lockorder....  mempool.cs has to be taken before cs_xval

File: src/txmempool.cpp [hunk 3]
```diff
@@ -292,7 +311,88 @@ bool CTxMemPool::CalculateMemPoolAncestors(const CTxMemPoolEntry &entry,
...
+        }
+        else if (totalSizeWithAncestors > limitAncestorSize)
+        {
+            errString =
+                strprintf(" %u exceeds ancestor size limit [limit: %u]", totalSizeWithAncestors, limitAncestorSize);
+            return false;
+        }
+
+        const setEntries &setMemPoolParents = GetMemPoolParents(stageit);
+        BOOST_FOREACH (const txiter &phash, setMemPoolParents)
```
Comment (sickpig) [inline src/txmempool.cpp]: if it is new  code why not using c++11 `for` loop rather than `BOOST_FOREACH`?
ofc this "complain" to all other cases where `BOOST_FOREACH` has been added in new code.

**Review by sickpig — DISMISSED** [review]

**Comment by sickpig** [comment]
@gandrewstone didn't mean to approve the PR (don't have the knowledge to do it). sorry.

File: src/thinblock.cpp (hunk 2 above)
Comment (gandrewstone) [reply src/thinblock.cpp:168]: It shouldn't because you can't get into a deadlock situation just by removing a lock.  You need to reverse the lock order.  Note that if this code locked mempool.cs again below cs_xval then removing the lock here would change the locking order.  So this might be the source of your confusion.  

Just to be sure that the checker does not incorrectly flag this, I ran the following code under a debug build and it works:
    {
         LOCK(cs_orphancache);
         READLOCK(mempool.cs);
         LOCK(cs_xval);
         printf("l1\n");
    }
    {
         LOCK(cs_orphancache);
         LOCK(cs_xval);
         printf("l2\n");
    }

File: qa/pull-tester/rpc-tests.py (hunk 1 above)
Comment (gandrewstone) [reply qa/pull-tester/rpc-tests.py:146]: true, I found this bug while working on this PR, and the fix is so small and simple...

File: src/thinblock.cpp (hunk 2 above)
Comment (ptschip) [reply src/thinblock.cpp:168]: There used to be a lock order constraint on these three locks, however, after looking through your code again, i see that you've removed this restriction (which is now possible) because of your read locks.  This is great!!!  I don't have any other issues with this PR...will approve now.

**Review by ptschip — APPROVED** [review]
